In [ ]:
from numpy import random
import numpy as np
def check_letters(number_ex,number_letters):
    count=0
    for i in range(number_ex): #проводим  эксперименты
        x = random.randint(number_letters)#выбираем случайное  число
        
        
        while(x==0):#выбираем значения больше 0
          x = random.randint(number_letters)

        #print(x)
        labels=np.array(list(range(1,x+1)))#составляем список случайных писем
        values=labels.copy()# копируем чтобы потом перемешать
        np.random.shuffle(values)
        j=1
        for j in range(x):
          if values[j]==labels[j]:
            count+=1
            break
    return 'the accuracy is for iterations {0} and letters {1}  iterations : {2}'.format(number_ex,number_letters,count/number_ex)

In [ ]:
print(check_letters(80000,50)) #т.е. для n=50, проведем 80000 экспериментов
print(check_letters(160000,50)) #увеличим количество экспериментов вдовое
print(check_letters(100000,100)) #т.е. для n=100, проведем 100000 экспериментов
print(check_letters(200000,100)) #увеличим количество экспериментов вдовое


the accuracy is for iterations 80000 and letters 50  iterations : 0.6367
the accuracy is for iterations 160000 and letters 50  iterations : 0.63871875
the accuracy is for iterations 100000 and letters 100  iterations : 0.63474
the accuracy is for iterations 200000 and letters 100  iterations : 0.63423


In [ ]:
#проверка
n=100 #возьмем 100 писем и адресов
#элементарные исходы, это всевозможные перестановки(рахмещения) n писем по n конвертам. т.е. если одно письмо точно попало в нужный конверт
#общее число таких перестановок это n!, а благоприятных событию "попало в свой конверт одно письмо" (n-1)!
#поэтому вероятность, что ровно одно письмо лежит в своем конверте: p=(n-1)!/n!=1/n  для любого из n писем.
#но у нас по условию задачи: хотя бы одно письмо, т.е. варианты 2,3,...k...n писем попавших в свой конверт нас тоже устроят. Для 3х писем попавших в нужные конверты p=(n-3)!/n!=1/n(n-1)(n-2)
#для n писем по своим конвертам вероятность равна p=1/n!, учтем еще сколькими способами можно выбрать k из n, т.е. положить ровно 3 письма в свои конверты
#подставляем в формулу это все т.е. P(хотя бы одно письмо)=n*1/n-C(2,n)*1/n(n-1)+C(3,n)*1/n(n-1)(n-2)-....+(-1)^(n-1)*1/n!=1-1/2!+1/3!-...+(-1)^(n-1)*1/n!
P=1
i=2
import math
for i in range(n):
  P=P+pow((-1),(i-1))/math.factorial(i)
print(P)

n=50
P=1
i=2
import math
for i in range(n):
  P=P+pow((-1),(i-1))/math.factorial(i)
print(P)


0.6321205588285578
0.6321205588285578


Получилось, что для n=100 вероятность равна 0.63, а для n=50 вероятность равна 0.63 с точностью в 2 знака после запятой. 


Вывод: 
Можно заметить, что с увеличением числа итераций в два раза, вероятность с точностью в 2 знака после запятой осталась такой же. Также с помощью формулы для классической вероятности была проведена проверка корректности вычислений и ответ совпал.